### DISCUSSION TITLE
$_com1 = "virtual" turns backspace input into "."

### DISCUSSION THREAD

**Discussion by bertrandom** [dosemu2/dosemu2#1983] [General]
This is difficult to reproduce, but I'm running a BBS door game, Legend of the Red Dragon (LORD) v4.00a and have it configured to use com port 1. I've set `$_com1 = "virtual"` in my `.dosemurc`. When sending the backspace character to the program, it registers it as backspace initially, but after a certain period of time or a certain number of characters sent back and forth, it turns into a `.` and stays that way for the rest of the session (all subsequent backspaces turn into periods).

I can try to make a simpler test case but I was wondering if anyone had any ideas what's going on here.

**Comment by stsp** [comment]
What output back-end is that, -t?
Instead of "virtual" you can try
`$_com1 = "exec rxvt -pty-fd 0"`
to get your game in a separate
rxvt window. See if the problem
is also there.

**Comment by bertrandom** [comment]
I wasn't able to get that working, but I tried this:
```
$_com1 = "exec cat - > /tmp/dosemu.com1"
```

and did `tail -F /tmp/dosemu.com1` in a different terminal and it didn't exhibit the same issues, it still handled input fine from the process but the backspace -> period bug never occurred.

**Reply by stsp** [reply]
If that works then your program is reading
the DOS console input, instead of getting
an input from com1. So I'd say your setup
bbs is misconfigured.
You can also try `$_com1 = "/dev/tty"` and
run dosemu with sdl backend. Then you
will have com1 at the terminal from where
you started dosemu, while the DOS window
will be separate.

**Reply by bertrandom** [reply]
How does `virtual` actually work then? With a door game, there's an admin stdin/stdout and there's also the communication over COM1.. are those both being routed to the terminal at the same time when `virtual` is used? Does it not accept input to `com1` when you select virtual?

**Reply by stsp** [reply]
It does accept input normally, but
obviously not when you set
`$_com1 = "exec cat - > /tmp/dosemu.com1"`
as this command doesn't produce any input.

> are those both being routed to the terminal at the same time when virtual is used?

Indeed, and that's the problem.
"virtual" is a rather bad hack.
You need to run dosemu in a
separate window and set
`$_com1 = "/dev/tty"` to at least
split the input from com1 and DOS
console.

**Reply by bertrandom** [reply]
Okay, I think I see what you're saying, when I change it from `virtual` to something else, the stdin is basically coming to the admin console, so it's not a great test of whether or not it's working in that case because I'm not sending keystrokes over COM1 (like I am in the `virtual` case).

I'm not actually running these in X windows, these are all running on a headless linux server. From my understanding, if `virtual` is used, then the admin stdin/stdout is basically discarded and the stdin/stdout for dosemu becomes the specified com port. So I need to try to direct COM1 to another device/file and write keystrokes to that to see if the same behavior exists..

**Reply by stsp** [reply]
There is a comredir command to redirect
the console to the com port, when in a
headless server you can't have another
window.
Another option is to use multiple virtual
consoles, for example by setting
`$_com1 = "/dev/tty5"` and switch to
tty5 with Alt-f5.

**Reply by stsp** [reply]
To use comredir, you first need to
connect 2 com ports with null-modem:
```
$_com1 = "nullmodem 2"
$_com2 = "nullmodem 1"
```
Then with comredir you can attach
the console to com2, and your game
will read it from com1.

**Reply by bertrandom** [reply]
> To use comredir, you first need to
> connect 2 com ports with null-modem:
> 
> $_com1 = "nullmodem 2"
> $_com2 = "nullmodem 1"
> Then with comredir you can attach
> the console to com2, and your game
> will read it from com1.

This seems to work but the admin screen is also writing to stdout - can you think of a way to suppress it?

**Reply by stsp** [reply]
`comredir -s`?
Once you get things to work, please
consider helping these guys:
http://wiki.synchro.net/howto:dosemu2
They seem to be unable to write the
proper howto.

**Comment by stsp** [comment]
I decided that this "virtual" hack can
be supported for a bit longer, as its
unclear when people migrate to an
alternatives.
So I applied the patch-set #1985 
Please let me know if it now works
any better.

**Comment by bertrandom** [comment]
I just installed it (`2.0~pre9-8003-dcc9a97d4+202304061203~ubuntu20.04.1`) but unfortunately I'm getting the same issue.. would I be helpful if I provide a ZIP with the C drive contents with everything configured so you can test it yourself?

**Reply by stsp** [reply]
This version does not have patches.
Maybe it wasn't yet rebuilt on LP.
I am not sure why you still haven't used
`comredir -s`.

**Reply by bertrandom** [reply]
Sorry, I tried `comredir -s` and while it did redirect the output of COM1 to stdout, it did not suppress the stdout output of the program, I think I forgot to mention that.

**Reply by bertrandom** [reply]
Okay, I can wait for the next build and try that.

**Reply by stsp** [reply]
Then we need to deal with that,
and that test-case would be interesting.
"virtual" is not interesting as being
legacy, though you'll still have a chance
to re-test it tomorrow.

**Comment by bertrandom** [comment]
BTW thanks so much for helping debug this.. a lot of the guides that I've read that suggest using `dosemu2` for BBS doors all suggest using `virtual` for the com port so I think it would be helpful to a lot of people (assuming the problem isn't just me).

**Comment by bertrandom** [comment]
I did some debugging last night with a small turbo pascal program - it seems like when virtual is activated:

1. The output of COM1 is sent to stdout
2. stdout from the program itself is disabled or sent nowhere (in my case the admin screen) - this is what I want, but just confirming.
3. stdin to dosemu is sent to the program

Is it true that there's no way to write to COM1 when `virtual` is used then? It just replaces the stdout in dosemu?

**Reply by stsp** [reply]
You can write to COM1, but DOS console
also gets the same input. This is a bug which
comredir doesn't have.

**Reply by bertrandom** [reply]
I imagine this is the overarching issue with `virtual` then if it sends it to both? Is that what your patch addresses?

You can see why in the use case of a BBS door this would be bad - the user could send a keystroke that the admin console would receive and process.

**Comment by bertrandom** [comment]
So.. I installed the new version `2.0~pre9-8005-60ebcae1c+202304071203~ubuntu20.04.1` and now whenever I press backspace it **always** echoes a `.`

**Comment by bertrandom** [comment]
Ok, I wrote a short program that reads and writes to the com port and your patch seems to improve the situation. It seems like when I send a backspace, the program receives an ASCII code of 127 (DEL). Shouldn't it receive an ASCII code of 8 (Backspace)?

**Comment by stsp** [comment]
Since you tested the patch, the last
reply in this thread is a due:
`-s` likely doesn't work for you because
you use ansi.sys and do not force the
dumb mode. So either disabling ansi
or running with `-dumb` should get `-s`
to work.
Ignore this once again, and don't bother
with postings.

**Reply by bertrandom** [reply]
I'm not loading ANSI.SYS in my CONFIG.SYS. I followed your instructions:

1. Connected the two comports will a null modem:
```
$_com1 = "nullmodem 2"
$_com2 = "nullmodem 1"
```

2. Wrote a batch file `C:\TEST.BAT` that runs `comredir` and then my test program:
```
@ECHO OFF
E:\DOSEMU\COMREDIR -s 2
C:\UTILS\COMECHO
```

3. Called dosmeu with the following flags:
```
dosemu -f dosemurc-null-modem -E "C:\\TEST" -dumb
```

This did suppress `stdout` as you suggested, but it seems to buffer the input. When I hit a character, the program won't register it until I hit enter.

---

On the other hand, I think your patch for `$_com1 = "virtual"` **did** work for addressing the issue where STDIN would be sent to the STDIN of the program rather than the virtual com port.

If I compare the 8003 and 8005 builds, the 8005 build now correctly sends STDIN to the virtual com port (without the buffering issues of `comredir`). Thank you for that.

I'm still having the issue that I mentioned before where hitting backspace sends ASCII code 127 (DEL) instead of ASCII code 8 (BS), but I can deal with that by rewriting the input upstream to me calling dosemu if we don't want to address it in dosemu.

I wrote two programs, one that reads STDIN and echoes the ASCII code of the character being pressed to STDOUT, and another that reads from com port 1 and echoes the ASCII code of the character received to com port 1 (as well as writing some initial data to STDOUT to test the suppression).

When I run the first program in dosemu, pressing backspace returns ASCII code 8. But when I use `$_com1 = "virtual"` and run the second program, pressing backspace returns ASCII code 127. I'm happy to provide both of these programs if they're useful to you.

**Reply by stsp** [reply]
Please try `-dumb -kt`.

**Reply by stsp** [reply]
Or actually `-td` instead of `-dumb`.

**Reply by bertrandom** [reply]
Yes, that worked! 🎉 And the ASCII code is returned as 8 when I hit backspace when using the null modem trick and comredir.

**Reply by stsp** [reply]
Thanks.
I wrote `comredir` very recently, and now
at least I know when/how it works in real
life, so I can document it properly.

As for the second part of the problem, I
added the patch that sets VERASE code
to 8, which should help.

**Reply by bertrandom** [reply]
Is there any way to disconnect the com port manually, perhaps by sending a signal to the dosemu process? I'd like to be able to simulate a modem disconnect. Thanks!

**Comment by stsp** [comment]
comredir recognizes ^z as disconnect,
although I dont think it can affect the
modem status. It can definitely affect
the line status, eg it can clear dtr, but
I dont remember if I implemented that. :)
Of course if you are going to insist on
virtual again, then you are on your own.
